### PR TITLE
Create order service

### PR DESCRIPTION
**Pull request by pmarsh-scottlogic**
Completed implementation of OrderService, including unit testing!

### CODE REVIEW

File: src/main/java/com/example/demo/matcher/models/Order.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/models/Order.java]: I suggest using this library for constructors its pretty good :) https://www.baeldung.com/intro-to-project-lombok it will automatically generates your getters, setters too 👍

File: src/main/java/com/example/demo/matcher/models/Order.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/models/Order.java]: Good idea overwriting the `toString` method.

File: src/main/java/com/example/demo/matcher/models/OrderbookItem.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/models/OrderbookItem.java]: Nice choosing the `Comparable` not `Comparator`

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/services/OrderService.java]: It looks nicer if you divide stream into different methods.
```java
        List<Order> filtered = orders.stream()
            .filter(order -> order.getAction() == action && Objects.equals(order.getAccountId(), accountId))
            .collect(Collections.toList());
```

File: src/main/java/com/example/demo/matcher/models/Order.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/models/Order.java]: Look at the `UUID` class, there is a method to get a random one on it.

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/services/OrderService.java]: You should "program to the interface", so return `List` instead of `ArrayList` (same for variable assignments), and only specify the type when you create an object ie `new WhateverList();`

**Review by nataliadombrowska — CHANGES_REQUESTED** [review]
Very good for your first attempt 👍 Just a couple of comments for now ill have a look over again once you've implemented those. Please don't merge until I approve.

File: src/main/java/com/example/demo/matcher/models/Order.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/models/Order.java]: What a cool library-  that'll shrink my files down some :)

File: src/main/java/com/example/demo/matcher/models/Order.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/models/Order.java]: Lombok does this too! Incredible

File: src/main/java/com/example/demo/matcher/models/OrderbookItem.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/models/OrderbookItem.java]: Credit goes to StackExchange :P

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/services/OrderService.java]: Yeah, the auto-formatting in IntelliJ is not a slick as VSCode. I'll make some of the long lines a bit shorter

File: src/main/java/com/example/demo/matcher/models/Order.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/models/Order.java]: Ah, nice and easy! I though I'd have to faff about with external libraries.

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/services/OrderService.java]: Okay dokey :) I guess that means that we can be more generic, and change which list implementation we're using down the line without having to refactor everything.

File: src/main/java/com/example/demo/matcher/models/Order.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/models/Order.java]: What's the difference between `id` and `accountId`?

File: src/main/java/com/example/demo/matcher/models/OrderbookItem.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/models/OrderbookItem.java:11]: Nice! This looks really clean now 👍

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/services/OrderService.java:32]: I don't think you have to call `this.get()` instead you can just do your stream on `orders` directly. I'm assuming the getter will be useful for other classes trying to access the orders from this service.

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/services/OrderService.java]: Would this work here instead: ` .collect(Collectors.toList())`

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (nataliadombrowska) [inline src/main/java/com/example/demo/matcher/services/OrderService.java]: In java it's best to use bigdecimal instead of float for monetary values. This is because you lose a bit of accuracy with float.

File: src/test/java/com/example/demo/OrderServiceTests.java
Comment (nataliadombrowska) [inline src/test/java/com/example/demo/OrderServiceTests.java:28]: Good job using `spy` I think you could even use `@Spy` annotation here too. Your preference though as it doesn't really make a difference.

File: src/test/java/com/example/demo/OrderServiceTests.java
Comment (nataliadombrowska) [inline src/test/java/com/example/demo/OrderServiceTests.java]: Not a blocker - might be worth pulling the orders that are duplicated across the tests in the `@BeforeEach`

**Review by nataliadombrowska — CHANGES_REQUESTED** [review]
Just a couple more comments. It's looking really good though 👍

File: src/main/java/com/example/demo/matcher/models/Order.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/models/Order.java]: id is the id for the order, and accountId is the id for the account. Should I rename it to orderId?

Really accountID should also be a UUID type - I'll get on that.

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/services/OrderService.java:32]: I'm using get so that I can mock the return value for testing. For example the test ItShouldGenerateAnOrderbookWithBuyAction(), I'm mocking the orderService.get() value on line 70

File: src/main/java/com/example/demo/matcher/models/Order.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/models/Order.java]: done

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/services/OrderService.java:32]: Is this good practise?

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/services/OrderService.java]: done

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (pmarsh-scottlogic) [reply src/main/java/com/example/demo/matcher/services/OrderService.java]: BigDecimals are so much worse to work with! But I see why it's important

File: src/test/java/com/example/demo/OrderServiceTests.java
Comment (pmarsh-scottlogic) [reply src/test/java/com/example/demo/OrderServiceTests.java:28]: I added the annotation next to the declaration of orderService for clarity, but I couldn't figure out how to instantiate it away from the declaration without using spy().

File: src/test/java/com/example/demo/OrderServiceTests.java
Comment (pmarsh-scottlogic) [reply src/test/java/com/example/demo/OrderServiceTests.java]: I've put them as returns from static methods instead, since there are different ones for different groups of tests

File: src/main/java/com/example/demo/matcher/services/OrderService.java
Comment (nataliadombrowska) [reply src/main/java/com/example/demo/matcher/services/OrderService.java:32]: Ideally we wouldn't do this but yeah for the mocking this makes it easier. I think this is fine to leave 👍

File: src/test/java/com/example/demo/OrderServiceTests.java
Comment (nataliadombrowska) [inline src/test/java/com/example/demo/OrderServiceTests.java:28]: I see what you're doing now. I was looking at this: https://www.baeldung.com/mockito-annotations when I reviewed this.

You could try getting rid of the `@Spy` annotation on line 23.

Then do something like this in the before each:
```java
@Spy
orderService = new OrderService();
```

I'm not sure if that works. Try it, if not we can leave it, again it doesn't really make a difference :)

**Review by nataliadombrowska — APPROVED** [review]
Good job 👍  no blockers - just added a comment about the @Spy annotation but happy for you to leave it 😄

**Review by nataliadombrowska — APPROVED** [review]

File: src/test/java/com/example/demo/OrderServiceTests.java
Comment (pmarsh-scottlogic) [reply src/test/java/com/example/demo/OrderServiceTests.java:28]: image

Comment (pmarsh-scottlogic) [reply src/test/java/com/example/demo/OrderServiceTests.java:28]: I'll get rid of the annotation for now. I had a little google and didn't find any quick answers

**Pull request merged**
